File: src/net/justinwhite/scoreit_p10/Game.java
/*
 Copyright (c) 2015 Justin White <jw@justinwhite.net>
 All rights reserved.

 Redistribution and use in source and binary forms, with or without
 modification, are permitted provided that the following conditions are met:

 1. Redistributions of source code must retain the above copyright notice,
 this list of conditions and the following disclaimer.

 2. Redistributions in binary form must reproduce the above copyright notice,
 this list of conditions and the following disclaimer in the documentation
 and/or other materials provided with the distribution.

 3. Neither the name of the copyright holder nor the names of its
 contributors may be used to endorse or promote products derived from this
 software without specific prior written permission.

 THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 POSSIBILITY OF SUCH DAMAGE.

 */

package net.justinwhite.scoreit_p10;

import java.util.*;

import static net.justinwhite.scoreit_p10.Util.println;

class Game {
    static Integer MIN_PLAYERS = 2;
    static Integer MAX_PLAYERS = 8;
    static Integer MAX_PHASE = 10;

    private final UUID id;
    private Integer numPlayers;
    private String name;
    private List<Player> players;
    private Map<String, Player> playerMap;
    private Player winner;

    public Game(Integer _numPlayers) {
        this.id = UUID.randomUUID();

        // sanity check number of players
        if (_numPlayers < MIN_PLAYERS) {
            _numPlayers = MIN_PLAYERS;
        } else if (_numPlayers > MAX_PLAYERS) {
            _numPlayers = MAX_PLAYERS;
        }

        numPlayers = 0;
        name = "";
        players = new ArrayList<Player>(_numPlayers);
        playerMap = new TreeMap<String, Player>();

        // add default players
        for (Integer i = 0; i < _numPlayers; i++) {
            addPlayer();
        }

        buildName();
    }

    public String toString() {
        return String.format("Game: %s\nUUID: %s\nPlayer count: %d\nPlayers: %s\nPlayerMap: %s",
                             getName(),
                             getID(),
                             getNumPlayers(),
                             // List<> and Map<> classes handle toString() themselves
                             players,
                             playerMap
        );
    }

    public UUID getID() {
        return id;
    }

    public void buildName() {
        name = "";
        for (Player p : players) {
            name += p.getInitials();
        }
    }

    public String getName() {
        return name;
    }

    public Integer getNumPlayers() {
        return numPlayers;
    }

    public List<Player> getPlayerList() {
        return players;
    }

    public Player getPlayer(Integer index) {
        return players.get(index);
    }

    public Player getPlayerByName(String _name) {
        return playerMap.get(_name);
    }

    public Boolean checkPlayer(String _name) {
        return playerMap.containsKey(_name);
    }

    public void addPlayer(String _name) {
        Player newPlayer = new Player();
        players.add(numPlayers++, newPlayer);
        playerMap.put(_name, newPlayer);
        newPlayer.setName(_name);
    }

    public void addPlayer() {
        addPlayer(String.format("Player %d", numPlayers + 1));
    }

    public void renamePlayer(String oldName, String newName) {
        Player p = playerMap.remove(oldName);
        if (p != null) {
            p.setName(newName);
            playerMap.put(newName, p);
        }
    }

    public String getScores() {
        String out = "";
        for (Player p : players) {
            out += String.format("%s: %4d Points", p.getName(), p.getScore());
            if (p.getPhase() > 0) {
                out += String.format(", Phase #%d completed\n", p.getPhase());
            } else {
                out += "\n";
            }
        }
        return out;
    }

    // TODO: handle mutiple winners: tie-break on score
    public Boolean hasWinner() {
        return winner != null;
    }

    public Player getWinner() {
        return (winner != null) ? winner : null;
    }

    public void setWinner(Player _winner) {
        winner = _winner;
    }

}

File: src/main/kotlin/net/justinwhite/scoremodel/Game.kt
/*
 * Copyright (c) 2015 Justin White <jw@justinwhite.net>
 *  All rights reserved.
 *
 *  Redistribution and use in source and binary forms, with or without
 *  modification, are permitted provided that the following conditions are met:
 *
 *  1. Redistributions of source code must retain the above copyright notice,
 *  this list of conditions and the following disclaimer.
 *
 *  2. Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 *  3. Neither the name of the copyright holder nor the names of its
 *  contributors may be used to endorse or promote products derived from this
 *  software without specific prior written permission.
 *
 *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 *  POSSIBILITY OF SUCH DAMAGE.
 */

package net.justinwhite.scoremodel

import java.text.SimpleDateFormat
import java.util.*

open class Game<T : Player>(private val playerClass: Class<T>, newNumPlayers: Int, _name: String?) {

    constructor(_playerClass: Class<T>, _numPlayers: Int)
    : this(_playerClass, _numPlayers, "")

    constructor(_playerClass: Class<T>, _name: String)
    : this(_playerClass, MIN_PLAYERS, _name)

    constructor(_playerClass: Class<T>)
    : this(_playerClass, MIN_PLAYERS, "")

    val ID: UUID = UUID.randomUUID()

    val playerList: MutableList<T> = ArrayList()

    private var _numPlayers: Int = 0
    var numPlayers: Int
        set(value) {
            var newNumPlayers: Int
            // bounds check number of players
            if (value < MIN_PLAYERS) {
                newNumPlayers = MIN_PLAYERS
            } else if (value > MAX_PLAYERS) {
                newNumPlayers = MAX_PLAYERS
            } else {
                newNumPlayers = value
            }
            // increasing count, add players
            if (newNumPlayers > numPlayers) {
                for (i in numPlayers..newNumPlayers - 1) {
                    addPlayer()
                }
            }
            // decreasing count, remove players
            if (numPlayers > newNumPlayers) {
                for (i in numPlayers downTo newNumPlayers + 1) {
                    removePlayer()
                }
            }
            // record new count
            _numPlayers = playerList.size
        }
        get() = _numPlayers

    var name = _name
        set(value) {
            field = if (value == null || value == "") {
                Companion.buildName()
            } else {
                value
            }
        }

    private var _winner: T? = null
    open public var winner: T?
        get() {
            checkWinner()
            return _winner
        }
        set(value) {
            _winner = value
        }

    init {
        numPlayers = newNumPlayers
    }


    override fun toString(): String {
        return "Game: %s\nUUID: %s\nPlayer count: %d\nPlayers: %s".format(name, ID, numPlayers, // List<> class handles toString()
                playerList)
    }

    // check if there is a player at the given index
    fun checkPlayer(index: Int): Boolean {
        return index >= 0 && index < numPlayers
    }

    // make a new Player [or subclass of player] using newInstance() and set it up for use
    private fun makePlayer(_name: String): T? {
        val newPlayer: T
        // to make a new Player [or subclass], use the class's instance factory
        try {
            newPlayer = playerClass.newInstance()
        } catch (e: InstantiationException) {
            // if the factory can't make a new object, something major is wrong
            e.printStackTrace()
            return null
        } catch (e: IllegalAccessException) {
            e.printStackTrace()
            return null
        }

        // successful instantiation, continue setting up the Player [or subclass]
        newPlayer.name = _name
        return newPlayer
    }

    // add a new player using _name
    @JvmOverloads fun addPlayer(_name: String = "Player " + (numPlayers + 1)): T? {
        if (_numPlayers < MAX_PLAYERS) {
            val newPlayer = makePlayer(_name)
            playerList.add(newPlayer as T)
            _numPlayers = playerList.size
            return newPlayer
        } else {
            return null
        }
    }

    // remove and return player specified by index. return null if index out of bounds
    @JvmOverloads fun removePlayer(index: Int? = numPlayers - 1): T? {
        if (_numPlayers > MIN_PLAYERS) {
            val oldPlayer = playerList.removeAt(index!!)
            _numPlayers = playerList.size
            return oldPlayer
        } else {
            return null
        }
    }

    // return player specified by index; first or last if index out of bounds
    fun getPlayer(_index: Int): T {
        var index = when {
            _index < 0 -> 0
            _index >= numPlayers -> numPlayers - 1
            else -> _index
        }
        return playerList[index]
    }

    // return an array of raw scores
    fun getScores(): List<Int> {
        val scores = ArrayList<Int>(playerList.size)
        for (p in playerList) {
            scores.add(p.score)
        }
        return scores
    }

    // return String of "Player1: Score1; Player2: Score2; etc"
    fun getScoresText(): String {
        var out = ""
        for (p in playerList) {
            out += "%s: %d Points; ".format(p.name, p.score)
        }
        return out
    }

    // tie-breaker is which ever player was added to the game first
    open fun checkWinner(): Boolean {
        val _players = ArrayList(playerList)
        Collections.sort(_players)
        _winner = _players[0]
        return _winner != null
    }

    companion object {
        const val MIN_PLAYERS = 2
        const val MAX_PLAYERS = 8

        fun buildName(): String {
            val format = "yyyy-MM-dd HH:mm"
            val sdf = SimpleDateFormat(format, Locale.US)
            return sdf.format(Date())
        }
    }
}
